### PR TITLE
Deferred fossil collection

### PR DESCRIPTION
**Pull request by Piccions**
Originally, fossil collection operations were executed synchronously after a new GVT was computed.
With this PR, fossil collection of a LP is deferred until we have to process a message for it. This measurably improves performance thanks to the improved cache friendliness.

Also, the memory footprint of the model exposed library has been halved with those changes:
- The pointer set by the model developer with the ```SetState()``` API is not saved anymore in the model memory allocator; that means that it can only be safely used during LP initialization

Also, some other little changes:
- Improved the handling of early remote anti messages, i.e. anti messages delivered before their positive counterpart
- Improved Doxygen documentation of the process module

### CODE REVIEW

**Review by alessandropellegrini — CHANGES_REQUESTED** [review]
Thanks a lot! Some minor rollbacks as we have discussed are needed before merging this one.

File: src/ROOT-Sim.h [hunk 1]
```diff
@@ -95,12 +95,14 @@ extern void *rs_realloc(void *ptr, size_t req_size);
 
 extern double Random(void);
 extern uint64_t RandomU64(void);
-extern double Expent(double mean);
-extern double Normal(void);
+extern double Expent(void);
+struct normal_deviates {
+	double d1, d2;
+};
```
[the remark this answers is not in the document]
Comment (Piccions) [reply src/ROOT-Sim.h]: Done!

File: src/gvt/gvt.c [hunk 2]
```diff
@@ -43,7 +43,7 @@ static __thread enum thread_phase thread_phase = thread_phase_idle;
 
 static timer_uint gvt_timer;
 static simtime_t reducing_p[MAX_THREADS];
-static __thread simtime_t current_gvt;
+static __thread simtime_t gvt_partial;
```
[the remark this answers is not in the document]
Comment (Piccions) [reply src/gvt/gvt.c]: Done!

File: src/lib/random/random.c [hunk 3]
```diff
@@ -147,38 +118,35 @@ double Gamma(unsigned ia)
...
 
 /**
- * Return the waiting time to the next event in a Poisson process of unit mean.
+ * @brief Return a random number according to an Exponential distribution with unit mean
+ * Corresponds to the waiting time to the next event in a Poisson process of unit mean.
  *
  * @return A random number
  */
-double Poisson(void)
+double Expent(void)
```
[the remark this answers is not in the document]
Comment (Piccions) [reply src/lib/random/random.c]: Done (inserted a macro in ROOT-Sim.h)

File: src/lp/lp.c [hunk 4]
```diff
@@ -129,23 +128,10 @@ void lp_fini(void)
 }
 
 /**
- * @brief Do housekeeping operations on the thread-locally hosted LPs after a fresh GVT
- * @param gvt the value of the freshly computed GVT
+ * @brief Set the LP simulation state main pointer
+ * @param state The state pointer to be passed to ProcessEvent() for the invoker LP
  */
-void lp_on_gvt(simtime_t gvt)
+void SetState(void *state)
```
[the remark this answers is not in the document]
Comment (Piccions) [reply src/lp/lp.c:138]: Done, the check is enabled only in the debug mode

File: src/gvt/fossil.c [hunk 5]
```diff
@@ -9,38 +9,42 @@
...
+static __thread simtime_t gvt_current;
+
+/**
+ * @brief Perform fossil collection operations at a given GVT
+ * @param current_gvt The value of the freshly computed GVT
+ */
+void fossil_on_gvt(simtime_t current_gvt)
+{
+	fossil_epoch_current += 1;
+	gvt_current = current_gvt;
```
[the remark this answers is not in the document]
Comment (Piccions) [reply src/gvt/fossil.c]: Fixed!

**Review by alessandropellegrini — APPROVED** [review]
I am happy with this PR, thanks a lot Andrea!
Just a little note: could you update the description in the initial message removing the reference to the Normal distribution?
I am asking this because it's automatically linked in the release notes.

Thanks!

**Pull request merged**